airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/io/PartitionedWriters.kt

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.destination.iceberg.v2.io
+
+import java.io.IOException
+import java.io.UncheckedIOException
+import org.apache.iceberg.FileFormat
+import org.apache.iceberg.PartitionKey
+import org.apache.iceberg.PartitionSpec
+import org.apache.iceberg.Schema
+import org.apache.iceberg.data.InternalRecordWrapper
+import org.apache.iceberg.data.Record
+import org.apache.iceberg.io.FileAppenderFactory
+import org.apache.iceberg.io.FileIO
+import org.apache.iceberg.io.OutputFileFactory
+import org.apache.iceberg.io.PartitionedWriter
+import org.apache.iceberg.util.Tasks
+
+/**
+ * [BaseDeltaTaskWriter] implementation that handles delta (insert, update, delete) writes to a
+ * partitioned Iceberg [org.apache.iceberg.Table].
+ */
+class PartitionedDeltaWriter(
+    spec: PartitionSpec,
+    format: FileFormat,
+    appenderFactory: FileAppenderFactory<Record>,
+    outputFileFactory: OutputFileFactory,
+    io: FileIO,
+    targetFileSize: Long,
+    schema: Schema,
+    identifierFieldIds: Set<Int>,
+) :
+    BaseDeltaTaskWriter(
+        spec,
+        format,
+        appenderFactory,
+        outputFileFactory,
+        io,
+        targetFileSize,
+        schema,
+        identifierFieldIds
+    ) {
+
+    private val partitionKey = PartitionKey(spec, schema)
+    private val writers = mutableMapOf<PartitionKey, RowDataDeltaWriter>()
```

Comment:
is each RowDataDeltaWriter holding nontrivial resources? (e.g. any sort of in-memory IO buffer)

b/c if we're partitioning on PK, there could be a lot of these writers, which could cause memory (or other) pressure, depending on what they're doing

Reply:
@edgao The `RowDataDeltaWriter` extends off of the Iceberg `BaseEqualityDeltaWriter` which does hold data about what has been inserted/deleted.  That all gets cleared out when the writer is closed.  The `IcebergStreamLoader` closes the writer within the `processRecords` method, so I guess the driving factor here will be whether that method gets batches of records or ALL of the records for a given stream, causing it to be long lived.  If it is the former, than we should be okay.